### PR TITLE
feat: Add React Query Utils helper

### LINKED ISSUE #4773: feat: Stable context during next navigation and thus fewer rerendering
### Describe the feature you'd like to request

It would be beneficial if the context provided by `withTrpc` that is accessible via `useContext` is stable when navigating in next. I have several places where i Use mutations in my code that trigger unnecessary rerenders on every url search param change which in my application is relatively frequent.

### Describe the solution you'd like to see

I would want the object generated at https://github.com/trpc/trpc/blob/main/packages/react-query/src/shared/hooks/createHooksInternal.tsx#L78-L241 to be memoized so it is stable.

### Describe alternate solutions

I could extract all TRPC code into its own component and memoize the callbacks so that this wrapping component is the only thing being rerendered and the child components only get callbacks that are momoized using useEventCallback or similar, this is however both inconvenient and error prone.

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4773">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4773/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### LINKED ISSUE #4974: feat: Use query helpers outside of Context
### Describe the feature you'd like to request

Like `useUtils` but can be accessed from outside of React context. Basically [this context value](https://github.com/trpc/trpc/blob/a98dc662d6ade1bcfdae09a274d508c9dc347235/packages/react-query/src/shared/hooks/createHooksInternal.tsx#L79) to be made available from a function that is called with a `client` and `queryClient`.

This would help us do prefetching of data at the router level, where context is not accessible. Having the full set of query utils would also be useful at this level for actions where we might want to perform cache invalidation after a mutation.

It's a [recommended pattern in the TanStack Query docs](https://tanstack.com/query/latest/docs/react/guides/prefetching#router-integration) for implementing prefetching and enabling fetch-as-you-render patterns with the now stable Suspense support.

### Describe the solution you'd like to see

A function that creates the same object as `useUtils`. It would take in the `client` and `queryClient` as parameters.

```ts
import { createQueryUtils } from '@trpc/react-query';

const utils = createQueryUtils({ client, queryClient });
await utils.post.all.prefetch();
```

Example usage in a React Router loader, a function which executes before React rendering and has no access to Context or hooks.

```tsx
import { createQueryUtils } from '@trpc/react-query';

const routes = [{
  path: ':postId',
  element: <Post />,
  loader: ({ params }) => {
    // Assume we have a way of getting the client instances into here depending on server/client.
    const utils = createQueryUtils({ client, queryClient });
    
    // Don't navigate until main content is ready.
    await utils.post.byId.prefetch(params.postId);

    // Start fetching comments early but don't block navigation.
    void utils.post.comments.prefetch(params.postId);
  },
}];
```

### Describe alternate solutions

`createServerSideHelpers` only works on the server, meaning we could only do this kind of prefetching for the initial SSR load. We can't perform prefetching or implement these same fetching patterns on the client.

It's more useful to have this at the router level so we can have the same prefetching logic working seamlessly on both server and client.

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/4974">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/4974/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### LINKED ISSUE #5129: feat: Query Client Utility Methods outside React Context
### Describe the feature you'd like to request

In my project I would like to be able to prefetch a query by using the `ensureData()` function in a react-router-loader to pre-populate the query client cache.

### Describe the solution you'd like to see

I would like to see a generic `utils` instead of `useUtils()` so that ensureData may be called outside of the react context but still be typed with tRPC and populate the QueryClient query cache.

### Describe alternate solutions

Currently in my project I'm defining a vanilla trpc client and using getQueryKey to find the key for the query I wish to pre-fetch. 
```ts
const result = await queryClient.ensureQueryData({
    queryKey: getQueryKey(trpc.auth.status),
    queryFn: () => trpcProxyClient.auth.status.query(),
  });
  ```

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5129">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5129/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by sebasptsch**
Closes #5129
Closes #4974 
Closes #4773 

## 🎯 Changes

Add typed react-query client utility methods for use outside React context.

This is well explained in #4974 

[RQ v4 SSR](https://tanstack.com/query/v4/docs/react/guides/ssr#using-nextjs)
[RQ v4 React Router](https://tanstack.com/query/v4/docs/react/examples/react/react-router)

Additionally the RQ NextJS docs declare a new QueryClient in `getStaticProps` so there isn't any cross-contamination between multiple users using SSR.
```tsx
import { dehydrate, QueryClient, useQuery } from '@tanstack/react-query'

export async function getStaticProps() {
  const queryClient = new QueryClient()

  await queryClient.prefetchQuery({ queryKey: ['posts'], queryFn: getPosts })

  return {
    props: {
      dehydratedState: dehydrate(queryClient),
    },
  }
}

function Posts() {
  // This useQuery could just as well happen in some deeper child to
  // the "Posts"-page, data will be available immediately either way
  const { data } = useQuery({ queryKey: ['posts'], queryFn: getPosts })

  // This query was not prefetched on the server and will not start
  // fetching until on the client, both patterns are fine to mix
  const { data: otherData } = useQuery({
    queryKey: ['posts-2'],
    queryFn: getPosts,
  })

  // ...
}
```

```tsx
const [queryClient] = React.useState(() => new QueryClient())

  return (
    <QueryClientProvider client={queryClient}>
      <Hydrate state={pageProps.dehydratedState}>
        <Component {...pageProps} />
      </Hydrate>
    </QueryClientProvider>
  )
  ```
(example from RQ SSR Docs)

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [x] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Comment by sebasptsch** [comment]
Merged from `next` into fork, bringing up to date. Ready for review.

**Review by KATT — CHANGES_REQUESTED** [review]
We're chatting in this thread on Discord: https://discord.com/channels/867764511159091230/1185712645777657978

Got some requests there, ping me there when you want us to re-review :)

File: packages/react-query/src/shared/proxy/utilsProxy.ts [hunk 1]
```diff
@@ -332,3 +336,48 @@ export function createReactQueryUtils<TRouter extends AnyRouter, TSSRContext>(
...
+        cancel: () => context.cancelQuery(queryKey, ...args),
+        setData: () => {
+          context.setQueryData(queryKey, args[0], args[1]);
+        },
+        setInfiniteData: () => {
+          context.setInfiniteQueryData(queryKey, args[0], args[1]);
+        },
+        getData: () => context.getQueryData(queryKey),
+        getInfiniteData: () => context.getInfiniteQueryData(queryKey),
+      };
```
Comment (juliusmarminge) [inline packages/react-query/src/shared/proxy/utilsProxy.ts:378]: we already have this somewhere, should be able to reuse it rigth?

Comment (juliusmarminge) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:378]: oh it's right above here

Comment (sebasptsch) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:378]: > oh it's right above here, this looks scary similar so can most likely get rid of one?
> 
> https://github.com/trpc/trpc/blob/5f66bfba30b32ecab6237cbd4402c63ff5d6ba2c/packages/react-query/src/shared/proxy/utilsProxy.ts#L287-L383

It looks like there's some SSR functionality embedded in the other one which is why I haven't just reused a bunch of stuff from the other utils proxy.

Comment (sebasptsch) [reply packages/react-query/src/shared/proxy/utilsProxy.ts]: @juliusmarminge I think I've made it as reusable as possible in the commit following this conversation.

File: packages/react-query/src/shared/hooks/createHooksInternal.tsx [hunk 2]
```diff
@@ -72,6 +67,16 @@ export function createRootHooks<
...
     const [ssrState, setSSRState] = useState<SSRState>(props.ssrState ?? false);
+
+    const queryFns = useMemo(
+      () =>
+        createUtilityFunctions({
+          client,
+          queryClient,
+        }),
+      [client, queryClient],
+    );
```
Comment (KATT) [inline packages/react-query/src/shared/hooks/createHooksInternal.tsx:78]: Could you memo the whole `value`? then we get #4773 for free

```ts
  const fns = useMemo(
    () => createTRPCQueryUtils({ queryClient, client }),
    [queryClient, client],
  );

  const value: ProviderContext = useMemo(
    () => ({
      ...fns,
      abortOnUnmount,
      queryClient,
      client,
      ssrContext: ssrContext ?? null,
      ssrState,
    }),
    [abortOnUnmount, client, queryClient, ssrContext, ssrState, fns],
  );
```

Comment (sebasptsch) [reply packages/react-query/src/shared/hooks/createHooksInternal.tsx]: Cool, done

**Review by KATT — APPROVED** [review]

**Review by KATT — APPROVED** [review]

**Pull request merged**